pandas/core/algorithms.py

```diff
@@ -395,6 +395,14 @@ def isin(comps, values):
                         " to isin(), you passed a [{values_type}]"
                         .format(values_type=type(values).__name__))
 
+    # GH 25507
```

Comment:
I would like you to integrate this with the other impl., meaning skip the set-ifying if its a set, but otherwise dispatch the actual isin operation. I suspect this will work only in some of the tests cases (but we aren't fully testing it here).

Reply:
I'm not sure what you mean by the other implementation or the actual isin operator. Do you mean let the rest of this function run but without the type cast? Or attempt to call comps.isin? It seems that would fail if comps was an ndarray.

Reply:
what i mean is you can skip the list-ifying if its a set, but don't actually do the comp in values. as I said I suspect this actually doesn't work (some tests are failing) and if we add sets to a fair number of tests they will fail. The reason is that these all must be the same types exactly e.g. comp is often a int while values maybe be an np.int. we already fully handle this path, so you need to fit in.

Reply:
The tests that failed are unrelated to my changes (conda connection issues on setup). I will add more tests for set values, but dispatching the current path of isin will destroy the performance gain this change intended to fix. 

I can't find any situations which cause this change to fail. The int vs np.int works fine, and I can create tests around such cases.

Reply:
you are creating a new path which makes the codes more complex ; we already have too many paths here

Reply:
Okay then feel free to close this pull request; the current path cannot support this performance change.

Reply:
as I said this needs to integrate with the existing isin tests. this needs to happen before this patch is considered.

Reply:
So the new patch works well for every test case except NaN types. The current paths ensure that an array containing NaN when compared against a collection with NaN in it returns True. However using Python sets NaN != NaN so it will return False (not sure of the reasoning, but I think current implementation goes against IEEE).

This means that `.isin({np.nan})` and `.isin([np.nan])` will have different results, and I don't know of a way to resolve this without introducing an O(n) check that destroys the performance fix.

Closing this pull request as I don't think we can implement this performance change without this regression.